### DISCUSSION TITLE
urpmi --no-verify is broken by rpm checking on its own with rpm-sequoia

### DISCUSSION THREAD

**Discussion by soig** [rpm-software-management/rpm#3203] [General]
We are hitting issues on Mageia with 4.20.x.
The first issue is that signatures being refused

This one seems to be on Mageia :

> error: Verifying a signature using certificate 00EDB89585B012A8916F0DF8B742FA8B80420F66 (Mageia Packages <[packages@mageia.org](mailto:packages@mageia.org)>):
>   1. Certificiate B742FA8B80420F66 invalid: certificate is not alive
>       because: The primary key is not live
>       because: Expired on 2012-03-13T12:10:11Z
>   2. Key B742FA8B80420F66 invalid: key is not alive
>       because: The primary key is not live
>       because: Expired on 2012-03-13T12:10:11Z

However this break "urpmi --no-verify" which used to behave like rpm --noverify
But now only rpm --nosignature can bypass this check, whereas rpm --no-verify used to do the same, which looks like regression ?

Previously, we used to check pkgs in urpmi but it looks like rpm does it on its own now too:
http://gitweb.mageia.org/software/rpm/urpmi/tree/urpm/main_loop.pm#n527
Which calls:
http://gitweb.mageia.org/software/rpm/urpmi/tree/urpm/main_loop.pm#n129
Which calls:
http://gitweb.mageia.org/software/rpm/urpmi/tree/urpm/signature.pm#n40
Which calls:
http://gitweb.mageia.org/software/rpm/perl-URPM/tree/URPM.xs#n3306
Which calls rpmReadPackageFile()

See attached logs showing urpmi output with both rpm-4.19 & 4.20
For me, there's 2 issues : 
- rpmlib < 4.20 not doing some checks
- rpmlib  redoing the checks performed by urpmi, and thus breaking urpmi when the user already says "yes, let's continue" after urpmi warned him/her
- same problem when using --no-verify which is the same as the user answering "y" above

Do we've to set RPMPROB_FILTER_VERIFY when calling rpmtsRun() ?
Would it impact urpmi running on older rpm versions (aka no more checks there) ?
Or should we kill that check (but then no more informative error in GUI?

Note that we all those upstream patches have been backported:
0001-Ensure-noarch-packages-don-t-get-debuginfo.patch
0001-Fix-noprep-regression-from-introducing-mkbuilddir.patch
0002-Drop-an-accidentally-added-duplicated-test.patch
0003-Make-build-in-place-much-less-of-a-hack-and-also-wor.patch
0001-Fix-incomplete-header-on-plain-src.rpm-build-modes-r.patch
0001-Hammer-in-no-debuginfo-for-noarch-packages-damn-it-r.patch
0001-Fix-regression-on-subpackage-debuginfo-RPMTAG_SOURCE.patch
0001-Fix-a-buildroot-regression-on-an-early-__spec_instal.patch

See you

[LOG.urpmi-rpm-4.19.txt](https://github.com/user-attachments/files/15545757/LOG.urpmi-rpm-4.19.txt)
[LOG.urpmi-rpm-4.20.txt](https://github.com/user-attachments/files/15545758/LOG.urpmi-rpm-4.20.txt)

**Comment by pmatilai** [comment]
The key management and all hasn't changed in rpm itself, this is rpm-sequoia being far, far stricter about what is accepted than the rpm internal pgp parser ever was. To go with rpm-sequoia, you'll need to keep your keys in shape, like not have keys expired 12 years ago.

The alternative is using the "traditional" PGP parser, it's still available only now split into separate repo (see INSTALL)

**Comment by soig** [comment]
Well the key is updated on media (=repos in urpmi).
So regular people who do upgrades with the installer will get the new key when removing old media, adding new media.
But I'm doing online updates for a decade using Cauldron (~ Rawhide for Fedora):

> diff -u KEY-*
> --- KEY-AVAIL2  2024-06-04 17:07:29.192787460 +0200
> +++ KEY-INST2   2024-06-04 17:07:37.793998092 +0200
> @@ -1,4 +1,4 @@
> -pub   rsa4096/0xB742FA8B80420F66 2011-02-07 [SCEA] **[expires: 2025-12-31]**
> +pub   rsa4096/0xB742FA8B80420F66 2011-02-07 [SCEA] **[expired: 2012-03-13]**
>        00EDB89585B012A8916F0DF8B742FA8B80420F66
>  uid                             Mageia Packages <[packages@mageia.org](mailto:packages@mageia.org)>
> -sig        0xB742FA8B80420F66 **2019-06-21**   [selfsig]
> +sig        0xB742FA8B80420F66 **2011-02-07**   [selfsig]

I guess I'm a special case 😅
But I may have to change urpmi.update so that it refresh the pubkey when it changes 😅

**Comment by soig** [comment]
Looks like even when adding new media, we don't re-import the key:
"pubkey 80420f66 already imported"
So I'll need to fix that.

Still @pmatilai, since urpmi checks pkgs first, should it set RPMPROB_FILTER_VERIFY when calling rpmtsRun() for real since at that point the user said yes to force ?

**Comment by pmatilai** [comment]
Yeah rpm doesn't have a meaningful way to update keys because back when I added the "is primary key id present", we didn't know anything about expiry, revokes, subkeys .. nothing. 
https://github.com/rpm-software-management/rpm/pull/3083 promises to fix that, until then the best you can do is try importing, if it already exists erase the key and try again. 

I don't know what urpmi --no-verify does, exactly. The story on rpm side is complicated as there are multiple related and partially overlapping mechanisms: the verify step during transactions is not a "signature check" per se, it's more of a payload validation step. Signatures can be used in there, but aren't by default. If urpmi's --no-verify is supposed to be a direct --nosignature counterpart then see 5ed632ae8a2a4a6de69a0d800340c9dc5d529ec0.

**Comment by soig** [comment]
So there's two issues, so I'll use 2 parts :

**Updating/reimporting a pubkey:**
Stupid question but : how can one remove a key from rpmdb (through the API, not using rpm -e) ?
we've rpmtsImportPubkey() but no remove or refresh API.
There's also rpmKeyringAddKey() but no RemoveKey().

Currently the URPM C binding uses rpmtsImportPubkey() https://gitweb.mageia.org/software/rpm/perl-URPM/tree/URPM.xs#n3363
(called from the perl interface : https://gitweb.mageia.org/software/rpm/perl-URPM/tree/URPM/Signature.pm#n82)

➡️ **_I guess I'll have to run a small transaction removing the offending key before importing._**
I don't see anything else ?
WDYT?

**Urpmi --no-verify**
By default (unless one uses --no-verify), urpmi first checks packages
http://gitweb.mageia.org/software/rpm/urpmi/tree/urpm/main_loop.pm#n527
before doing the actual installation:
http://gitweb.mageia.org/software/rpm/urpmi/tree/urpm/main_loop.pm#n544

The actual installation basically means we run a series of transactions here http://gitweb.mageia.org/software/rpm/urpmi/tree/urpm/install.pm#n352
my $trans = $db->create_transaction;
$trans->check
$trans->order
$trans->run
Which are defined in URPM binding : https://gitweb.mageia.org/software/rpm/perl-URPM/tree/URPM.xs#n2770 : respectively rpmtsCheck(), rpmtsOrder() and rpmtsRun()

But before installing, as said above, unless using --no-verify, we'll do some tests : http://gitweb.mageia.org/software/rpm/urpmi/tree/urpm/main_loop.pm#n129
which really happens here: http://gitweb.mageia.org/software/rpm/urpmi/tree/urpm/signature.pm#n40
For each package we call verify_signature() which is there in the URPM binding: https://gitweb.mageia.org/software/rpm/perl-URPM/tree/URPM.xs#n3306 where we use rpmReadPackageFile() to check each package file.

So at this point we already checked the pkgs for signatures.
If in interactive mode (either CLI or GUI) we ask the user if he wants to abort and force the installation.
Previously, this worked fine, but now rpmlib checks again the pkgs after we already did it and it doesn't know that the user actually said "frak this, install anyway".

So I think that in urpmi context, we want to prevent rpmlib to check again the pkgs since we already did it.
➡️ _**Does passing RPMPROB_FILTER_VERIFY would be enough to achieve this**_ when calling rpmtsRun()?

**Comment by pmatilai** [comment]
> I guess I'll have to run a small transaction removing the offending key before importing.

Yes, there's nothing better at the moment.

> Does passing RPMPROB_FILTER_VERIFY would be enough to achieve this when calling rpmtsRun()?

Maybe, but again that verify step is NOT about signature verification as such, it could be package whose (header) signature is perfectly valid but truncated payload, or such. For "frak the signatures", IIRC you want to set rpmtsSetVfyFlags() to same as rpmtsSetVSFlags().

Note that letting verify do its job has more subtle side-effects too: installations only show as verified in the auditing log if rpm itself verified it.

**Comment by soig** [comment]
> > I guess I'll have to run a small transaction removing the offending key before importing.
> 
> Yes, there's nothing better at the moment.

I've a working patch for that.
Though that's not ideal b/c when adding a whole set of media/repos (core, nonfree, tainted -- each one existing as regular, updates, updates_testing), that mean adding/removing the same key for all.
Is there a way to get the expiry date with rpm API?

BTW other packagers are throwing stones at mea, saying there's a difference between an expired key and a revoked
key : a signature that was valid yesterday doesn't suddenly become invalid today,
just because the key expired!"
 
> > Does passing RPMPROB_FILTER_VERIFY would be enough to achieve this when calling rpmtsRun()?
> 
> Maybe, but again that verify step is NOT about signature verification as such, it could be package whose (header) signature is perfectly valid but truncated payload, or such. For "frak the signatures", IIRC you want to set rpmtsSetVfyFlags() to same as rpmtsSetVSFlags().
> 
> Note that letting verify do its job has more subtle side-effects too: installations only show as verified in the auditing log if rpm itself verified it.

That would only be set up if the urpmi verify pass failed and the user said "go on anyway" or if he sets up the no-verify option in /etc/urpmi/urpmi.cfg ot used the --no-verify option, so in regular cases that won't be an issue

**Comment by pmatilai** [comment]
> BTW other packagers are throwing stones at mea, saying there's a difference between an expired key and a revoked
key : a signature that was valid yesterday doesn't suddenly become invalid today,
just because the key expired!"

It gets all a bit filosophical. If it doesn't come invalid at that point, then when does it? And if it doesn't, what's the point of expiry in the first place? It shouldn't be possible to create new signatures with an expired key for one, but I guess expiry should be considered wrt the time of signing instead of "now" I suppose. I kinda thought that's how Sequoia does it, but not sure (tagging @nwalfield for possible comments). To evaluate what actually happens one needs to look at both the key and the signature, which we don't have linked here AFAICS.

**Reply by nwalfield** [reply]
Sorry for the late reply.  If a certificate expires (at say `t2`) that doesn't mean that signatures made prior to the expiry (e.g., at `t1`, where `t1 < t2`) are suddenly invalid.  If that is happening, that is a bug, and I'd appreciate a report.

**Comment by soig** [comment]
Yeah but aborting on expired key when we don't have an API to refresh the key yet is an issue.
We should first add such an API, then later abort on expired keys.
As we say in French, "c'est mettre la charrue avant les boeufs" (=~ "it's putting the cart before the horse")😉
At least checking against signing date rather than now would be better in the current state of the rpms

**Comment by pmatilai** [comment]
Oh, no disagreement there. This is a case where basically everybody got caught with their pants down: in the first 20+ years of rpm's existence nobody gave much thought to expiry and all that, but then in recent years (even before sequoia) people started running into their ancient SHA1 based keys suddenly stopping to work, first due to FIPS and then more generally distros tried to push for disabling obsoleted crypto but disabling SHA1 got postponed and postponed and postponed for things like google-chrome, and then rpm-sequoia was introduced and pushing those people more strongly to update their signing systems. And when those parties finally updated their keys we realize there's no mechanism for them to update those keys at all. :see_no_evil:

**Comment by bcornec** [comment]
@soig, I just found this threa as I was looking for an error in updating my mageia cauldron docker container image :-(

After doing urpmi.update -a
and trying to install build deps for the tellico package, I had the following:
``
urpmi SPECS/tellico.spec
To satisfy dependencies, the following packages are going to be installed:
  Package                        Version      Release       Arch    
(medium "Core Release (mga1)")
  binutils                       2.43.1       2.mga10       x86_64  
  bm                             3.9          1.mga10       noarch  
  cmake-rpm-macros               9            9.mga9        noarch  
  elfutils                       0.192        4.mga10       x86_64  
  gcc                            14.2.0       2.mga10       x86_64  
  gcc-c++                        14.2.0       2.mga10       x86_64  
  gcc-cpp                        14.2.0       2.mga10       x86_64  
  gdb-headless                   15.2         1.mga10       x86_64  
  gdb-minimal                    15.2         1.mga10       x86_64  
  glibc                          2.40         1.mga10       x86_64  
  glibc-devel                    2.40         1.mga10       x86_64  
  gobject-introspection          1.82.0       1.mga10       x86_64  
  lib64asm1                      0.192        4.mga10       x86_64  
  lib64debuginfod1               0.192        4.mga10       x86_64  
  lib64dw1                       0.192        4.mga10       x86_64  
  lib64elf1                      0.192        4.mga10       x86_64  
  lib64mpdec4                    4.0.0        1.mga10       x86_64  
  lib64nsl3                      2.0.1        1.mga10       x86_64  
  lib64openblas-serial0          0.3.28       1.mga10       x86_64  (recommended)
  lib64openssl3                  3.3.2        2.mga10       x86_64  
  lib64pcre2posix3               10.44        1.mga10       x86_64  
  lib64python3.12                3.12.8       1.mga10       x86_64  
  lib64python3.12-stdlib         3.12.8       1.mga10       x86_64  
  lib64rpm10                     4.20.0       1.mga10       x86_64  
  lib64rpmbuild10                4.20.0       1.mga10       x86_64  
  lib64rpmsign10                 4.20.0       1.mga10       x86_64  
  lib64xcrypt-devel              4.4.37       1.mga10       x86_64  
  lib64xcrypt1                   4.4.37       1.mga10       x86_64  
  libgfortran5                   14.2.0       2.mga10       x86_64  (recommended)
  libgomp-devel                  14.2.0       2.mga10       x86_64  
  libgomp1                       14.2.0       2.mga10       x86_64  
  libquadmath0                   14.2.0       2.mga10       x86_64  (recommended)
  libstdc++-devel                14.2.0       2.mga10       x86_64  
  libstdc++6                     14.2.0       2.mga10       x86_64  
  mgarepo                        1.14.6       2.mga10       noarch  
  openblas                       0.3.28       1.mga10       x86_64  (recommended)
  openssl                        3.3.2        2.mga10       x86_64  
  pcre2-tools                    10.44        1.mga10       x86_64  
  perl                           5.40.0       3.mga10       x86_64  
  perl-File-Sync                 0.110.0      39.mga10      x86_64  
  perl-Filesys-Df                0.920.0      40.mga10      x86_64  
  perl-List-MoreUtils-XS         0.430        12.mga10      x86_64  
  perl-Locale-gettext            1.70.0       22.mga10      x86_64  
  perl-URPM                      5.225        4.mga10       x86_64  
  perl-XML-LibXML                2.21.0       8.mga10       x86_64  
  perl-base                      5.40.0       3.mga10       x86_64  
  pyproject-rpm-macros           1.16.3       1.mga10       noarch  
  pyproject-srpm-macros          1.16.3       1.mga10       noarch  
  python3                        3.12.8       1.mga10       x86_64  
  python3-dateutil               2.9.0.post0  1.mga10       noarch  (recommended)
  python3-httplib2               0.22.0       1.mga10       noarch  
  python3-numpy                  2.1.3        3.mga10       x86_64  (recommended)
  python3-packaging              24.0         1.mga10       noarch  
  python3-pandas                 2.2.2        2.mga10       x86_64  (recommended)
  python3-pip                    24.3.1       1.mga10       noarch  
  python3-pygments               2.17.2       1.mga10       noarch  
  python3-pyparsing              3.1.2        1.mga10       noarch  
  python3-pytz                   2024.1       1.mga10       noarch  (recommended)
  python3-rpm                    4.20.0       1.mga10       x86_64  
  python3-setuptools             75.6.0       1.mga10       noarch  
  python3-six                    1.16.0       6.mga10       noarch  
  python3-tqdm                   4.67.1       2.mga10       noarch  
  python3-tzdata                 2023.3       2.mga10       noarch  (recommended)
  python3-wheel                  0.43.0       1.mga10       noarch  
  rpm                            4.20.0       1.mga10       x86_64  
  rpm-build                      4.20.0       1.mga10       x86_64  
  rpm-mageia-setup               2.82         1.mga10       x86_64  
  rpm-mageia-setup-build         2.82         1.mga10       x86_64  
  rpm-sequoia                    1.7.0        1.mga10       x86_64  
  rust-srpm-macros               26.3         2.mga10       noarch  
260MB of additional disk space will be used.
158MB of packages will be retrieved.
Proceed with the installation of the 70 packages? (Y/n) y

[... download phase ....]

installing lib64rpmbuild10-4.20.0-1.mga10.x86_64.rpm lib64python3.12-stdlib-3.12.8-1.mga10.x86_64.rpm perl-XML-LibXML-2.21.0-8.mga10.x86_64.rpm python3-httplib2-0.22.0-1.mga10.noarch.rpm libgomp1-14.2.0-2.mga10.x86_64.rpm rpm-4.20.0-1.mga10.x86_64.rpm python3-packaging-24.0-1.mga10.noarch.rpm python3-tzdata-2023.3-2.mga10.noarch.rpm python3-pytz-2024.1-1.mga10.noarch.rpm pyproject-srpm-macros-1.16.3-1.mga10.noarch.rpm lib64xcrypt-devel-4.4.37-1.mga10.x86_64.rpm python3-dateutil-2.9.0.post0-1.mga10.noarch.rpm pcre2-tools-10.44-1.mga10.x86_64.rpm rpm-mageia-setup-build-2.82-1.mga10.x86_64.rpm python3-wheel-0.43.0-1.mga10.noarch.rpm mgarepo-1.14.6-2.mga10.noarch.rpm lib64rpmsign10-4.20.0-1.mga10.x86_64.rpm lib64rpm10-4.20.0-1.mga10.x86_64.rpm perl-base-5.40.0-3.mga10.x86_64.rpm lib64dw1-0.192-4.mga10.x86_64.rpm libgomp-devel-14.2.0-2.mga10.x86_64.rpm gcc-14.2.0-2.mga10.x86_64.rpm lib64debuginfod1-0.192-4.mga10.x86_64.rpm cmake-rpm-macros-9-9.mga9.noarch.rpm openssl-3.3.2-2.mga10.x86_64.rpm libgfortran5-14.2.0-2.mga10.x86_64.rpm python3-pandas-2.2.2-2.mga10.x86_64.rpm glibc-devel-2.40-1.mga10.x86_64.rpm binutils-2.43.1-2.mga10.x86_64.rpm python3-tqdm-4.67.1-2.mga10.noarch.rpm rust-srpm-macros-26.3-2.mga10.noarch.rpm python3-pygments-2.17.2-1.mga10.noarch.rpm lib64openssl3-3.3.2-2.mga10.x86_64.rpm perl-URPM-5.225-4.mga10.x86_64.rpm gcc-c++-14.2.0-2.mga10.x86_64.rpm perl-List-MoreUtils-XS-0.430-12.mga10.x86_64.rpm pyproject-rpm-macros-1.16.3-1.mga10.noarch.rpm perl-Filesys-Df-0.920.0-40.mga10.x86_64.rpm lib64mpdec4-4.0.0-1.mga10.x86_64.rpm elfutils-0.192-4.mga10.x86_64.rpm rpm-mageia-setup-2.82-1.mga10.x86_64.rpm python3-six-1.16.0-6.mga10.noarch.rpm bm-3.9-1.mga10.noarch.rpm lib64xcrypt1-4.4.37-1.mga10.x86_64.rpm glibc-2.40-1.mga10.x86_64.rpm python3-rpm-4.20.0-1.mga10.x86_64.rpm python3-numpy-2.1.3-3.mga10.x86_64.rpm openblas-0.3.28-1.mga10.x86_64.rpm python3-pip-24.3.1-1.mga10.noarch.rpm rpm-build-4.20.0-1.mga10.x86_64.rpm lib64pcre2posix3-10.44-1.mga10.x86_64.rpm perl-5.40.0-3.mga10.x86_64.rpm python3-pyparsing-3.1.2-1.mga10.noarch.rpm lib64python3.12-3.12.8-1.mga10.x86_64.rpm python3-setuptools-75.6.0-1.mga10.noarch.rpm perl-Locale-gettext-1.70.0-22.mga10.x86_64.rpm python3-3.12.8-1.mga10.x86_64.rpm perl-File-Sync-0.110.0-39.mga10.x86_64.rpm rpm-sequoia-1.7.0-1.mga10.x86_64.rpm gdb-minimal-15.2-1.mga10.x86_64.rpm libquadmath0-14.2.0-2.mga10.x86_64.rpm lib64asm1-0.192-4.mga10.x86_64.rpm lib64nsl3-2.0.1-1.mga10.x86_64.rpm lib64openblas-serial0-0.3.28-1.mga10.x86_64.rpm gcc-cpp-14.2.0-2.mga10.x86_64.rpm libstdc++6-14.2.0-2.mga10.x86_64.rpm gobject-introspection-1.82.0-1.mga10.x86_64.rpm lib64elf1-0.192-4.mga10.x86_64.rpm libstdc++-devel-14.2.0-2.mga10.x86_64.rpm gdb-headless-15.2-1.mga10.x86_64.rpm from /var/cache/urpmi/rpms
Preparing...                     #############################################
     1/70: glibc                 #############################################
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down
     2/70: lib64xcrypt1          #############################################
     3/70: perl-base             #############################################
     4/70: lib64elf1             #############################################
     5/70: lib64dw1              #############################################
     6/70: lib64debuginfod1      #############################################
     7/70: lib64openssl3         #############################################
     8/70: perl                  #############################################
     9/70: rpm-mageia-setup      #############################################
    10/70: rpm-sequoia           #############################################
    11/70: lib64rpm10            #############################################
    12/70: rpm                   #############################################
    13/70: lib64rpmsign10        #############################################
    14/70: lib64asm1             #############################################
    15/70: lib64xcrypt-devel     #############################################
    16/70: glibc-devel           #############################################
    17/70: lib64mpdec4           #############################################
    18/70: lib64pcre2posix3      #############################################
    19/70: pcre2-tools           #############################################
    20/70: lib64nsl3             #############################################
    21/70: python3               #############################################
    22/70: lib64python3.12-stdlib
                                 #############################################
    23/70: lib64python3.12       #############################################
    24/70: python3-setuptools    #############################################
    25/70: gobject-introspection #############################################
    26/70: python3-packaging     #############################################
    27/70: python3-tzdata        #############################################
    28/70: python3-pytz          #############################################
    29/70: python3-wheel         #############################################
    30/70: python3-six           #############################################
    31/70: python3-dateutil      #############################################
    32/70: python3-pip           #############################################
    33/70: python3-pyparsing     #############################################
    34/70: python3-httplib2      #############################################
    35/70: gcc-cpp               #############################################
    36/70: openblas              #############################################
    37/70: rust-srpm-macros      #############################################
    38/70: libstdc++6            #############################################
    39/70: binutils              #############################################
    40/70: libstdc++-devel       #############################################
    41/70: gcc                   #############################################
    42/70: libgomp1              #############################################
    43/70: lib64rpmbuild10       #############################################
    44/70: python3-rpm           #############################################
    45/70: gdb-headless          #############################################
    46/70: libgomp-devel         #############################################
    47/70: gcc-c++               #############################################
    48/70: elfutils              #############################################
    49/70: gdb-minimal           #############################################
    50/70: pyproject-srpm-macros #############################################
    51/70: cmake-rpm-macros      #############################################
    52/70: rpm-mageia-setup-build
                                 #############################################
    53/70: rpm-build             #############################################
    54/70: pyproject-rpm-macros  #############################################
    55/70: libquadmath0          #############################################
    56/70: bm                    #############################################
    57/70: perl-URPM             #############################################
    58/70: python3-pygments      #############################################
    59/70: perl-List-MoreUtils-XS
                                 #############################################
    60/70: openssl               #############################################
    61/70: perl-XML-LibXML       #############################################
Warning: program compiled against libxml 212 using older 209
Warning: XML::LibXML compiled against libxml2 21209, but runtime libxml2 is older 20914
    62/70: perl-Filesys-Df       #############################################
    63/70: perl-Locale-gettext   #############################################
    64/70: perl-File-Sync        #############################################
    65/70: libgfortran5          #############################################
    66/70: lib64openblas-serial0 #############################################
    67/70: python3-numpy         #############################################
    68/70: python3-pandas        #############################################
    69/70: python3-tqdm          #############################################
    70/70: mgarepo               #############################################

[ ... removal phase ...]
  
You should restart your computer for glibc
restarting urpmi
defaulting to --buildrequires
[ ... tons of packages to install ...]
2.8GB of additional disk space will be used.
714MB of packages will be retrieved.
Proceed with the installation of the 1333 packages? (Y/n) y
[ ... download phase ...]
error: Verifying a signature using certificate 00EDB89585B012A8916F0DF8B742FA8B80420F66 (Mageia Packages <packages@mageia.org>):
  1. Certificate B742FA8B80420F66 invalid: policy violation
      because: No binding signature at time 2025-01-15T18:36:25Z
      because: Policy rejected non-revocation signature (PositiveCertification) requiring second pre-image resistance
      because: SHA1 is not considered secure since 2023-02-01T00:00:00Z
  2. Certificate B742FA8B80420F66 invalid: policy violation
      because: No binding signature at time 2025-01-20T10:30:04Z
      because: Policy rejected non-revocation signature (PositiveCertification) requiring second pre-image resistance
      because: SHA1 is not considered secure since 2023-02-01T00:00:00Z
warning: /var/cache/urpmi/rpms/lib64mount1-2.40.4-1.mga10.x86_64.rpm: Header V4 RSA/SHA256 Signature, key ID 80420f66: NOTTRUSTED
error: Verifying a signature using certificate 00EDB89585B012A8916F0DF8B742FA8B80420F66 (Mageia Packages <packages@mageia.org>):
  1. Certificate B742FA8B80420F66 invalid: policy violation
      because: No binding signature at time 2022-03-18T20:37:16Z
      because: Policy rejected non-revocation signature (PositiveCertification) requiring second pre-image resistance
      because: SHA1 is not considered secure since 2023-02-01T00:00:00Z
  2. Certificate B742FA8B80420F66 invalid: policy violation
      because: No binding signature at time 2025-01-20T10:30:04Z
      because: Policy rejected non-revocation signature (PositiveCertification) requiring second pre-image resistance
      because: SHA1 is not considered secure since 2023-02-01T00:00:00Z

going on for all the remaining packages.

Have you found a workaround for this ?
